File: src/main/java/br/com/pnp/compiler/model/construct/statement/WhileStatement.java
package br.com.pnp.compiler.model.construct.statement;

import br.com.pnp.compiler.model.construct.Block;
import br.com.pnp.compiler.model.construct.type.Type;
import br.com.pnp.compiler.model.construct.type.primitives.PrimitiveType;
import br.com.pnp.compiler.model.expression.Expression;

public class WhileStatement extends Statement {
    private Block block;
    private Expression condition;

    public WhileStatement(Expression condition) {
        this.block = new StatementBlock();
        this.condition = condition;
    }

    @Override
    public Type getType() {
        return PrimitiveType.Booleano;
    }

    public Block getWhileBlock() {
        return this.block;
    }

    public Expression getCondition() {
        return this.condition;
    }
}

File: src/main/kotlin/br/com/pnp/model/construct/statement/WhileStatement.kt
package br.com.pnp.model.construct.statement

import br.com.pnp.model.construct.Block
import br.com.pnp.model.construct.Variable
import br.com.pnp.model.construct.type.primitive.PrimitiveType
import br.com.pnp.model.expression.Expression

class WhileStatement(condition: Expression) : Statement {
    val block: Block = StatementBlock()
    var condition: Expression = Variable.literalBoolean(false)
        set(value) {
            if (!PrimitiveType.boolean.isTypeOf(value.type)) {
                throw IllegalArgumentException(
                        """
                        Type mismatch in ${this.javaClass.simpleName} constructor:
                        Required: ${PrimitiveType.boolean}
                        Found: ${value.type}
                    """.trimIndent())
            }
            field = value
        }

    init {
        this.condition = condition
    }
}
